ethcore/src/state/mod.rs

```diff
@@ -940,20 +940,71 @@ impl<B: Backend> State<B> {
 	}
 
 	/// Populate a PodAccount map from this state.
-	pub fn to_pod(&self) -> PodState {
+	fn to_pod_cache(&self) -> PodState {
 		assert!(self.checkpoints.borrow().is_empty());
 		// TODO: handle database rather than just the cache.
 		// will need fat db.
-		PodState::from(self.cache.borrow().iter().fold(BTreeMap::new(), |mut m, (add, opt)| {
+		PodState::from(self.to_pod_cache_inner())
+	}
+
+	#[inline]
+	fn to_pod_cache_inner(&self) -> BTreeMap<Address, PodAccount> {
+		self.cache.borrow().iter().fold(BTreeMap::new(), |mut m, (add, opt)| {
 			if let Some(ref acc) = opt.account {
 				m.insert(add.clone(), PodAccount::from_account(acc));
 			}
 			m
-		}))
+		})
+	}
+
+
+	#[cfg(feature="to-pod-full")]
+	/// Populate a PodAccount map from this state.
+	/// Warning this is not for real time use.
+	/// Use of this method requires FatDB mode to be able 
+	/// to iterate on accounts.
+	pub fn to_pod_full(&self) -> PodState {
+		use ethereum_types::H160;
+
+		assert!(self.checkpoints.borrow().is_empty());
+		assert!(self.factories.trie.is_fat());
+
+		let mut result = self.to_pod_cache_inner();
+
+		let trie = match self.factories.trie.readonly(self.db.as_hashdb(), &self.root) {
+			Ok(trie) => trie,
+			_ => {
+				trace!(target: "state", "to_pod_full: Couldn't open the DB");
+				return PodState::from(result);
+			}
+		};
+
+		let iter = match trie.iter() {
+			Ok(iter) => iter,
+			_ => return PodState::from(result),
+		};
+
+		let from_rlp = |b: &[u8]| Account::from_rlp(b).expect("decoding db value failed");
+
+		for item in iter {
+			if let Ok((addr, dbval)) = item {
+				result.entry(Address::from_slice(&addr)).or_insert_with(|| {
+					let mut account = from_rlp(&dbval[..]);
+					if account.code_size().is_none() {
+						let addr_hash = account.address_hash(&H160::from(&addr[..]));
```

Comment:
Sorry I missed this -- this will also not completely work -- for a fresh account object, we would miss all storage items! And I do believe we have plenty of tests that deal with storage items. So I think we would also need to fix this.

We do already have logics in `cli::blockchain::execute_export_state` that do this. You can refer to its logic, and I would recommend one of the two ways to deal with this.

#### Refactor `Client::list_accounts` and `Client::list_storage`

Currently they only work for real client, but not EVM test client, but I believe it would be possible to extend this interface, and change the `count` argument to accept `Option`. With this, we change the signature to `to_pod_full(&self, accounts: HashMap<Address, Vec<H256>>)`, where for the hashmap, keys are account addresses, and values are existing storage items.

Within this function, call `State::require` for all addresses, and call `State::storage_at` for all storage items. After that, we can use `State::to_pod_cache` safely to get all addresses.

#### Simpler without Refactoring

Simply use the same logic as `Client::list_accounts` and `Client::list_storage`, but copy them to `State`, gather `HashMap<Address, Vec<H256>>` there. And use `State::require`, `State::storage_at`, and `State::to_pod_cache`.